### PR TITLE
RFC: add OS.Lifecycle module

### PR DESCRIPTION
**Pull request by talex5**
Allows unikernels to handle shutdown requests.

### CODE REVIEW

**Comment by djs55** [comment]
I bet @jonludlam would be interested

File: xen/lib/lifecycle.mli [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+val await_shutdown : unit -> [`Poweroff | `Reboot] Lwt.t
```
Comment (avsm) [inline xen/lib/lifecycle.mli]: It's not quite clear from the documentation how this is ever meant to return. If it resolves, then the domain has disappeared, right?

**Comment by jonludlam** [comment]
Looks very useful! I think we should advertise the features in xenstore - see https://github.com/mirage/xen/blob/master/docs/misc/xenstore-paths.markdown#controlfeature-poweroff--01-w

Having said that, I think xapi currently notices that you have the key set, but doesn't actually behave any differently - it always expects PV VMs to obey writes to /control/shutdown. We should stick to the documented spec though :-)

Do you have any plans to look at any other things to report back via xenstore mentioned in the doc? e.g. the IP address of the interfaces?

File: xen/lib/lifecycle.mli (hunk 1 above)
Comment (talex5) [inline xen/lib/lifecycle.mli]: Well, it does says "when the domain is *asked* to shut down" - perhaps it could be renamed to `await_shutdown_request`.

**Comment by talex5** [comment]
@jonludlam IP stuff is already handled by `mirage-net-xen`.

I did think about advertising the features. However, we don't really know what the unikernel wants to support, so maybe it should set things up itself if it has a preference?

I wonder if we should simplify the interface to return just `unit Lwt.t`. Is the distinction between poweroff and reboot useful for anything?

**Comment by jonludlam** [comment]
@talex5 won't you need to know what was asked for so you can pass it to OS.Sched.shutdown? Also, perhaps we could write the 'feature-shutdown' at the point the await_shutdown is called? Presumably the way this will be called is as a long-running background thread when the unikernel first starts up, so the key will be there if the toolstack decides to go looking for it.

**Comment by talex5** [comment]
@avsm: I've renamed `await_shutdown` to `await_shutdown_request`.

@jonludlam: good point about needing to know about poweroff vs reboot. I've added `?can_poweroff` and `?can_reboot` flags too to set the features. Can't test it as Qubes doesn't give me permission to write there anyway.

**Pull request merged**